### PR TITLE
ESLint: error on missing lint ignore comments

### PR DESCRIPTION
**Pull request by fungairino**
## What does this PR do?

- Starts to fix the newly enabled rule for enforcing explanations on eslint ignores
- I updated the new-cap rule so we don't have to ignore it a bunch
- There is one functional change I pointed out.

102 errors remaining

### CODE REVIEW

File: src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts [hunk 1]
```diff
@@ -114,16 +114,19 @@ async function setIntegrationDependencyVars(
...
-    // eslint-disable-next-line no-await-in-loop
-    const serviceContext = await makeServiceContextFromDependencies([
-      integrationDependency,
-    ]);
-    contextVars.setExistenceFromValues({
-      source: `${KnownSources.SERVICE}:${integrationDependency.integrationId}`,
-      values: serviceContext,
-    });
-  }
+  await Promise.all(
```
Comment (fungairino) [inline src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts:117]: This is the only functional change in the PR. This should be safe (and more efficient)

File: .eslintrc.js [hunk 2]
```diff
@@ -30,6 +30,12 @@ module.exports = {
     "@shopify/react-hooks-strict-return": "error",
     "@shopify/prefer-module-scope-constants": "error",
     "@shopify/jest/no-snapshots": "warn",
+    "new-cap": [
+      "error",
+      {
+        capIsNewExceptionPattern: "(TEST_|INTERNAL_|HACK_)",
```
Comment (twschiller) [inline .eslintrc.js:36]: 💯

File: src/mods/useMods.test.ts [hunk 3]
```diff
@@ -61,7 +61,7 @@ describe("useMods", () => {
     const wrapper = renderHook(() => useMods(), {
       setupRedux(dispatch) {
         dispatch(
-          // eslint-disable-next-line new-cap -- unsave
+          // eslint-disable-next-line new-cap -- unsafe
```
Comment (twschiller) [inline src/mods/useMods.test.ts:64]: Add to the ignore list for new-cap?

File: src/auth/authUtils.ts [hunk 4]
```diff
@@ -60,8 +60,8 @@ export function selectUserDataUpdate({
 
   return {
     // TODO: Review Required/Partial in Me type
-    // https://github.com/pixiebrix/pixiebrix-extension/issues/7725
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion, @typescript-eslint/no-unnecessary-type-assertion
+    //  https://github.com/pixiebrix/pixiebrix-extension/issues/7725
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion, @typescript-eslint/no-unnecessary-type-assertion -- email is required
```
Comment (twschiller) [inline src/auth/authUtils.ts:64]: For `!` comments, I'd expect to see an explanation of how we know it's non null

Comment (fungairino) [reply src/auth/authUtils.ts:64]: I believe this one in particular will be fixed soon by @BLoe 's work.

File: src/mods/useMods.test.ts (hunk 3 above)
Comment (fungairino) [reply src/mods/useMods.test.ts:64]: sure!

File: src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts (hunk 1 above)
Comment (fregante) [reply src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts:117]: We also have an `asyncForEach` helper so you don't have to use a `map` that doesn't return anything

File: src/analysis/analysisVisitors/varAnalysis/parseTemplateVariables.ts [hunk 5]
```diff
@@ -30,7 +30,7 @@ const VARIABLE_START_INDEX = Symbol("#Variable_start_index");
 
 export interface VariableOptions {
   type?: string;
-  // eslint-disable-next-line @typescript-eslint/no-use-before-define
+  // eslint-disable-next-line @typescript-eslint/no-use-before-define -- define type interface first
```
Comment (fregante) [inline src/analysis/analysisVisitors/varAnalysis/parseTemplateVariables.ts:33]:
```suggestion
  // eslint-disable-next-line @typescript-eslint/no-use-before-define -- Cyclic dependency
```

File: src/auth/authUtils.ts [hunk 6]
```diff
@@ -60,8 +60,9 @@ export function selectUserDataUpdate({
 
   return {
     // TODO: Review Required/Partial in Me type
-    // https://github.com/pixiebrix/pixiebrix-extension/issues/7725
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion, @typescript-eslint/no-unnecessary-type-assertion
+    //  https://github.com/pixiebrix/pixiebrix-extension/issues/7725
+    /* eslint-disable-next-line @typescript-eslint/no-non-null-assertion, @typescript-eslint/no-unnecessary-type-assertion
```
Comment (fregante) [inline src/auth/authUtils.ts:64]: I always assumed "comment on top" was supported. It turns out it isn't yet :(

- https://github.com/mysticatea/eslint-plugin-eslint-comments/issues/55


I think it's supported for `@ts-*` comments though

File: .eslintrc.js [hunk 7]
```diff
@@ -30,6 +30,12 @@ module.exports = {
     "@shopify/react-hooks-strict-return": "error",
     "@shopify/prefer-module-scope-constants": "error",
     "@shopify/jest/no-snapshots": "warn",
+    "new-cap": [
+      "error",
+      {
+        capIsNewExceptionPattern: "(TEST_|INTERNAL_|HACK_|UNSAFE_)",
+      },
+    ],
```
Comment (fregante) [inline .eslintrc.js:38]: 🙌

File: src/background/setToolbarIconFromTheme.test.ts [hunk 8]
```diff
@@ -51,7 +51,7 @@ describe("setToolbarIconFromTheme", () => {
   // @ts-expect-error -- No need to mock the whole class for the test
   global.Image = class {
     src = "";
-    // eslint-disable-next-line no-restricted-syntax
+    // eslint-disable-next-line no-restricted-syntax -- This is a mock
     decode = jest.fn().mockResolvedValue(undefined);
```
Comment (fregante) [inline src/background/setToolbarIconFromTheme.test.ts:55]: This comment shouldn't be there, the rule is specifically about mocks

Screenshot 5

```suggestion
    decode = jest.fn();
```

File: src/components/documentBuilder/preview/DocumentPreview.test.tsx [hunk 9]
```diff
@@ -87,43 +87,42 @@ describe("Add new element", () => {
...
-    // eslint-disable-next-line testing-library/no-node-access
+    // eslint-disable-next-line testing-library/no-node-access -- There's a row in the container and a column in the row.
     containerElement.children[0].children[0].children[0] = listElement;
 
     const { container } = renderDocumentPreview(containerElement);
 
     // Select a dropdown inside a Col in List and open it
     await userEvent.click(
-      // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
+      // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access -- see test's TODO comment
```
Comment (fregante) [inline src/components/documentBuilder/preview/DocumentPreview.test.tsx:97]: Nit: A single `/* eslint-disable testing-library/no-container, testing-library/no-node-access */` might be best

File: src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts (hunk 1 above)
Comment (grahamlangford) [reply src/analysis/analysisVisitors/varAnalysis/varAnalysis.ts:117]: I remember that helper, but it looks like it no longer exists?

**Review by grahamlangford — APPROVED** [review]

**Pull request merged**